src/framegrab/grabber.py

```
@@ -512,6 +511,79 @@ def _set_cv2_resolution(self) -> None:
+            if not success:
+                # Treat the failure to set FOURCC has an exception if the user explicitly requested a fourcc, otherwise just log it
+                message = f"Failed to set FOURCC to '{new_fourcc}' for camera '{self.config.name}'. The camera might not support this setting."
+                if new_fourcc_is_user_provided:
+                    raise RuntimeError(message)
+                else:
+                    logger.warning(message)
+
+    def _set_cv2_buffersize(self) -> None:
+        """Set buffer size from the config (2 when streaming, else 1).
+
+        Do nothing if unchanged. Raise RuntimeError if setting fails.
+        """
+        video_stream = self.config.video_stream
+
+        # Determine desired buffer size based on video_stream setting
+        if video_stream:
+            desired_buffer_size = 2  # Larger buffer for better FPS when streaming
+        else:
+            desired_buffer_size = 1  # Small buffer to always get most recent frame
+
+        # Check current buffer size
+        current_buffer_size = int(self.capture.get(cv2.CAP_PROP_BUFFERSIZE))
+
+        if desired_buffer_size != current_buffer_size:
+            success = self.capture.set(cv2.CAP_PROP_BUFFERSIZE, desired_buffer_size)
+            if not success:
+                logger.warning(
+                    f"Failed to set buffer size to {desired_buffer_size} for camera '{self.config.name}. The camera might not support this setting.'"
+                )
+
+    def _set_cv2_fps(self) -> None:
+        """Set FPS from the config.
+
+        Do nothing if FPS is None or unchanged. Raise RuntimeError if setting fails.
+        """
+        new_fps = self.config.fps
+
+        # Check if the caller set an FPS
+        new_fps_is_user_provided = new_fps is not None
+
+        # Set to a default FPS if the user provided no FPS
+        new_fps = new_fps if new_fps is not None else DEFAULT_FPS
```

The defaulting logic can all be handled in the pedantic config

---

I did this because I wanted to differentiate between the user explicitly setting a value which happens to be the same as the default, and the default being used because the user provided no value.

In the former case, I want to raise an exception if the FPS the user wanted cannot be set. 

In the later case, I want to log a warning if the default FPS cannot be applied, because some hardware might not support setting an FPS. 

To my knowledge, Pydantic doesn't have a way to do this.